File: src/services/ProfileService.test.ts
/**
 * Profile Service Unit Tests
 *
 * Tests for ProfileService functionality including:
 * - Profile data storage and retrieval
 * - Validation logic
 * - Error handling
 * - Memoization and caching
 */

import { describe, it, expect, beforeEach, afterEach, vi } from "vitest";
import { ProfileService } from "./ProfileService";
import type { UserProfile, ProfileFormData } from "../types/profile";
import { ProfileStorageKey } from "../types/profile";

// Mock localStorage globally
const localStorageMock = {
  getItem: vi.fn(),
  setItem: vi.fn(),
  removeItem: vi.fn(),
  clear: vi.fn(),
  length: 0,
  key: vi.fn(),
} as any;

// Mock console methods to avoid test output pollution
const consoleErrorMock = vi.fn();
const consoleWarnMock = vi.fn();

// Set up global mocks (use window.localStorage in jsdom environment)
Object.defineProperty(window, "localStorage", {
  value: localStorageMock,
  writable: true,
  configurable: true,
});

Object.defineProperty(global.console, "error", {
  value: consoleErrorMock,
  writable: true,
});

Object.defineProperty(global.console, "warn", {
  value: consoleWarnMock,
  writable: true,
});

describe("ProfileService", () => {
  let profileService: ProfileService;

  beforeEach(() => {
    profileService = new ProfileService();
    vi.clearAllMocks();
    localStorageMock.clear();

    // Restore our custom localStorage mock (tests/setup.ts overwrites it in global beforeEach)
    Object.defineProperty(window, "localStorage", {
      value: localStorageMock,
      writable: true,
      configurable: true,
    });
  });

  afterEach(() => {
    // Clean up any cached data
    profileService.clear();
  });

  describe("get()", () => {
    it("should return null when no profile data exists", () => {
      localStorageMock.getItem.mockReturnValue(null);

      const result = profileService.get();

      expect(result).toBeNull();
    });

    it("should return profile data when it exists", () => {
      localStorageMock.getItem.mockImplementation((key: string) => {
        switch (key) {
          case ProfileStorageKey.FIRST_NAME:
            return "John";
          case ProfileStorageKey.LAST_NAME:
            return "Doe";
          case ProfileStorageKey.EMAIL:
            return "john.doe@example.com";
          case ProfileStorageKey.PHONE:
            return "+1234567890";
          default:
            return null;
        }
      });

      const result = profileService.get();

      expect(result).toEqual({
        firstName: "John",
        lastName: "Doe",
        email: "john.doe@example.com",
        phone: "+1234567890",
      });
    });

    it("should return profile data with only email", () => {
      localStorageMock.getItem.mockImplementation((key: string) => {
        switch (key) {
          case ProfileStorageKey.EMAIL:
            return "john@example.com";
          default:
            return null;
        }
      });

      const result = profileService.get();

      expect(result).toEqual({
        firstName: "",
        lastName: "",
        email: "john@example.com",
        phone: undefined,
      });
    });

    it("should handle localStorage errors gracefully", () => {
      localStorageMock.getItem.mockImplementation(() => {
        throw new Error("localStorage error");
      });

      const result = profileService.get();

      expect(result).toBeNull();
      expect(consoleErrorMock).toHaveBeenCalledWith(
        "[ProfileService] Error retrieving profile:",
        expect.any(Error),
      );
    });
  });

  describe("update()", () => {
    it("should successfully update profile data", async () => {
      // Mock no existing data
      localStorageMock.getItem.mockReturnValue(null);
      localStorageMock.setItem.mockImplementation(() => {});

      const profileUpdate: Partial<UserProfile> = {
        firstName: "John",
        lastName: "Doe",
        email: "john.doe@example.com",
      };

      const result = await profileService.update(profileUpdate);

      expect(result.success).toBe(true);
      expect(result.message).toBe("Profile updated successfully");
      expect(result.updatedFields).toEqual({
        firstName: "John",
        lastName: "Doe",
        email: "john.doe@example.com",
        phone: undefined,
      });

      expect(localStorageMock.setItem).toHaveBeenCalledWith(
        ProfileStorageKey.FIRST_NAME,
        "John",
      );
      expect(localStorageMock.setItem).toHaveBeenCalledWith(
        ProfileStorageKey.LAST_NAME,
        "Doe",
      );
      expect(localStorageMock.setItem).toHaveBeenCalledWith(
        ProfileStorageKey.EMAIL,
        "john.doe@example.com",
      );
    });

    it("should validate profile data before updating", async () => {
      const invalidProfile: Partial<UserProfile> = {
        email: "invalid-email",
      };

      const result = await profileService.update(invalidProfile);

      expect(result.success).toBe(false);
      expect(result.message).toContain("Profile validation failed");
      expect(localStorageMock.setItem).not.toHaveBeenCalled();
    });

    it("should merge with existing profile data", async () => {
      // Mock existing profile
      localStorageMock.getItem.mockImplementation((key: string) => {
        switch (key) {
          case ProfileStorageKey.FIRST_NAME:
            return "Jane";
          case ProfileStorageKey.LAST_NAME:
            return "Smith";
          case ProfileStorageKey.EMAIL:
            return "jane.smith@example.com";
          default:
            return null;
        }
      });

      localStorageMock.setItem.mockImplementation(() => {});

      const profileUpdate: Partial<UserProfile> = {
        firstName: "John", // Override existing
        phone: "+1234567890", // Add new
      };

      const result = await profileService.update(profileUpdate);

      expect(result.success).toBe(true);
      expect(result.updatedFields).toEqual({
        firstName: "John", // Updated
        lastName: "Smith", // Preserved
        email: "jane.smith@example.com", // Preserved
        phone: "+1234567890", // Added
      });
    });
  });

  describe("validate()", () => {
    it("should validate email format", () => {
      const validProfile: Partial<ProfileFormData> = {
        email: "test@example.com",
      };

      const invalidProfile: Partial<ProfileFormData> = {
        email: "invalid-email",
      };

      expect(profileService.validate(validProfile).isValid).toBe(true);
      expect(profileService.validate(invalidProfile).isValid).toBe(false);
      expect(profileService.validate(invalidProfile).errors.email).toBe(
        "Please enter a valid email address",
      );
    });

    it("should validate phone number format", () => {
      const validProfile: Partial<ProfileFormData> = {
        phone: "+1234567890",
      };

      const invalidProfile: Partial<ProfileFormData> = {
        phone: "invalid-phone",
      };

      expect(profileService.validate(validProfile).isValid).toBe(true);
      expect(profileService.validate(invalidProfile).isValid).toBe(false);
      expect(profileService.validate(invalidProfile).errors.phone).toBe(
        "Please enter a valid phone number",
      );
    });

    it("should validate name character restrictions", () => {
      const validProfile: Partial<ProfileFormData> = {
        firstName: "John-Paul",
        lastName: "O'Connor",
      };

      const invalidProfile: Partial<ProfileFormData> = {
        firstName: "John@123",
      };

      expect(profileService.validate(validProfile).isValid).toBe(true);
      expect(profileService.validate(invalidProfile).isValid).toBe(false);
      expect(profileService.validate(invalidProfile).errors.firstName).toBe(
        "First name contains invalid characters",
      );
    });

    it("should return detailed error messages", () => {
      const invalidProfile: Partial<ProfileFormData> = {
        email: "invalid",
        phone: "abc",
        firstName: "John@123",
      };

      const result = profileService.validate(invalidProfile);

      expect(result.isValid).toBe(false);
      expect(Object.values(result.errors).filter(Boolean)).toHaveLength(3);
    });
  });

  describe("clear()", () => {
    it("should clear all profile data from localStorage", () => {
      profileService.clear();

      expect(localStorageMock.removeItem).toHaveBeenCalledWith(
        ProfileStorageKey.FIRST_NAME,
      );
      expect(localStorageMock.removeItem).toHaveBeenCalledWith(
        ProfileStorageKey.LAST_NAME,
      );
      expect(localStorageMock.removeItem).toHaveBeenCalledWith(
        ProfileStorageKey.FULL_NAME,
      );
      expect(localStorageMock.removeItem).toHaveBeenCalledWith(
        ProfileStorageKey.EMAIL,
      );
      expect(localStorageMock.removeItem).toHaveBeenCalledWith(
        ProfileStorageKey.PHONE,
      );
    });

    it("should handle localStorage errors gracefully", () => {
      localStorageMock.removeItem.mockImplementation(() => {
        throw new Error("Remove error");
      });

      expect(() => profileService.clear()).not.toThrow();
      expect(consoleErrorMock).toHaveBeenCalled();
    });
  });

  describe("getExtended()", () => {
    it("should return null when no profile exists", () => {
      localStorageMock.getItem.mockReturnValue(null);

      const result = profileService.getExtended();

      expect(result).toBeNull();
    });

    it("should return extended profile with computed fields", () => {
      localStorageMock.getItem.mockImplementation((key: string) => {
        switch (key) {
          case ProfileStorageKey.FIRST_NAME:
            return "John";
          case ProfileStorageKey.LAST_NAME:
            return "Doe";
          case ProfileStorageKey.EMAIL:
            return "john.doe@example.com";
          default:
            return null;
        }
      });

      const result = profileService.getExtended();

      expect(result).toEqual({
        firstName: "John",
        lastName: "Doe",
        email: "john.doe@example.com",
        phone: undefined,
        fullName: "John Doe",
        initials: "JD",
      });
    });

    it("should handle single name initials", () => {
      localStorageMock.getItem.mockImplementation((key: string) => {
        switch (key) {
          case ProfileStorageKey.FIRST_NAME:
            return "John";
          case ProfileStorageKey.EMAIL:
            return "john@example.com";
          default:
            return null;
        }
      });

      const result = profileService.getExtended();

      expect(result?.initials).toBe("J");
    });
  });

  describe("formDataToProfile() and profileToFormData()", () => {
    it("should convert between form data and profile data", () => {
      const formData: ProfileFormData = {
        firstName: "John",
        lastName: "Doe",
        email: "john@example.com",
        phone: "+1234567890",
      };

      const profile = profileService.formDataToProfile(formData);
      const convertedBack = profileService.profileToFormData(profile);

      expect(profile).toEqual({
        firstName: "John",
        lastName: "Doe",
        email: "john@example.com",
        phone: "+1234567890",
      });

      expect(convertedBack).toEqual(formData);
    });

    it("should handle undefined phone in profile to form conversion", () => {
      const profile: UserProfile = {
        firstName: "John",
        lastName: "Doe",
        email: "john@example.com",
        phone: undefined,
      };

      const formData = profileService.profileToFormData(profile);

      expect(formData.phone).toBe("");
    });
  });

  describe("caching behavior", () => {
    it("should cache extended profile results", () => {
      localStorageMock.getItem.mockImplementation((key: string) => {
        switch (key) {
          case ProfileStorageKey.FIRST_NAME:
            return "John";
          case ProfileStorageKey.LAST_NAME:
            return "Doe";
          case ProfileStorageKey.EMAIL:
            return "john@example.com";
          default:
            return null;
        }
      });

      // First call should compute and cache
      const result1 = profileService.getExtended();
      expect(result1?.fullName).toBe("John Doe");

      // Second call should use cache
      const result2 = profileService.getExtended();
      expect(result2).toBe(result1); // Same reference

      // Verify localStorage was only called once per field
      expect(localStorageMock.getItem).toHaveBeenCalledTimes(4); // firstName, lastName, email, phone
    });

    it("should invalidate cache when profile is updated", async () => {
      localStorageMock.getItem.mockReturnValue(null);
      localStorageMock.setItem.mockImplementation(() => {});

      // Get extended profile (should cache)
      const result1 = profileService.getExtended();
      expect(result1).toBeNull();

      // Update profile (should invalidate cache)
      await profileService.update({
        firstName: "John",
        email: "john@example.com",
      });

      // Get extended profile again (should recompute)
      localStorageMock.getItem.mockImplementation((key: string) => {
        switch (key) {
          case ProfileStorageKey.FIRST_NAME:
            return "John";
          case ProfileStorageKey.EMAIL:
            return "john@example.com";
          default:
            return null;
        }
      });

      const result2 = profileService.getExtended();
      expect(result2?.firstName).toBe("John");
    });
  });
});

File: backend/services/profile_service.test.py
"""
Unit tests for ProfileService.
Tests profile CRUD operations, validation, caching, and encryption.
"""

import pytest
import time
from unittest.mock import Mock, MagicMock
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from datetime import datetime

from backend.models.base import Base
from backend.models.profile import UserProfile
from backend.services.profile_service import (
    ProfileService,
    UserProfileData,
    ProfileValidationResult,
    ProfileUpdateResult,
    ExtendedUserProfileData,
    ProfileFormData,
)
from backend.services.encryption_service import EncryptionService


# Test fixtures
@pytest.fixture
def db_session():
    """Create in-memory SQLite database for testing."""
    engine = create_engine("sqlite:///:memory:")
    Base.metadata.create_all(engine)
    Session = sessionmaker(bind=engine)
    session = Session()

    # Create initial profile row (id=1)
    profile = UserProfile(
        id=1,
        name="Legal User",
        first_name=None,
        last_name=None,
        email=None,
        phone=None
    )
    session.add(profile)
    session.commit()

    yield session
    session.close()


@pytest.fixture
def encryption_service():
    """Mock encryption service."""
    service = Mock(spec=EncryptionService)
    service.encrypt = Mock(side_effect=lambda x: f"encrypted_{x}")
    service.decrypt = Mock(side_effect=lambda x: x.replace("encrypted_", ""))
    return service


@pytest.fixture
def audit_logger():
    """Mock audit logger."""
    logger = Mock()
    logger.log = Mock()
    return logger


@pytest.fixture
def profile_service(db_session, encryption_service, audit_logger):
    """Create ProfileService instance."""
    return ProfileService(
        db=db_session,
        encryption_service=encryption_service,
        audit_logger=audit_logger
    )


# Test cases
class TestProfileServiceGet:
    """Test get() method."""

    @pytest.mark.asyncio
    async def test_get_empty_profile(self, profile_service):
        """Test getting profile when no data exists."""
        result = await profile_service.get()
        assert result is None

    @pytest.mark.asyncio
    async def test_get_existing_profile(self, profile_service, db_session):
        """Test getting existing profile data."""
        # Create profile with data
        profile = db_session.query(UserProfile).filter(UserProfile.id == 1).first()
        profile.first_name = "John"
        profile.last_name = "Doe"
        profile.email = "encrypted_john@example.com"
        profile.phone = "encrypted_+1234567890"
        db_session.commit()

        result = await profile_service.get()

        assert result is not None
        assert result.firstName == "John"
        assert result.lastName == "Doe"
        assert result.email == "john@example.com"  # Decrypted
        assert result.phone == "+1234567890"  # Decrypted

    @pytest.mark.asyncio
    async def test_get_without_encryption_service(self, db_session, audit_logger):
        """Test getting profile without encryption service."""
        service = ProfileService(db=db_session, encryption_service=None, audit_logger=audit_logger)

        profile = db_session.query(UserProfile).filter(UserProfile.id == 1).first()
        profile.first_name = "Jane"
        profile.last_name = "Smith"
        profile.email = "jane@example.com"
        db_session.commit()

        result = await service.get()

        assert result is not None
        assert result.firstName == "Jane"
        assert result.email == "jane@example.com"  # Not decrypted


class TestProfileServiceUpdate:
    """Test update() method."""

    @pytest.mark.asyncio
    async def test_update_new_profile(self, profile_service, db_session, encryption_service):
        """Test creating new profile data."""
        profile_data = {
            "firstName": "Alice",
            "lastName": "Johnson",
            "email": "alice@example.com",
            "phone": "+9876543210"
        }

        result = await profile_service.update(profile_data)

        assert result.success is True
        assert result.message == "Profile updated successfully"
        assert result.updatedFields is not None
        assert result.updatedFields.firstName == "Alice"

        # Verify database
        db_profile = db_session.query(UserProfile).filter(UserProfile.id == 1).first()
        assert db_profile.first_name == "Alice"
        assert db_profile.last_name == "Johnson"
        assert db_profile.name == "Alice Johnson"
        assert db_profile.email == "encrypted_alice@example.com"
        assert db_profile.phone == "encrypted_+9876543210"

    @pytest.mark.asyncio
    async def test_update_partial_data(self, profile_service, db_session):
        """Test updating only some fields."""
        # Create initial profile
        profile = db_session.query(UserProfile).filter(UserProfile.id == 1).first()
        profile.first_name = "Bob"
        profile.last_name = "Smith"
        profile.email = "encrypted_bob@example.com"
        db_session.commit()

        # Update only email
        result = await profile_service.update({"email": "newemail@example.com"})

        assert result.success is True

        # Verify other fields unchanged
        db_profile = db_session.query(UserProfile).filter(UserProfile.id == 1).first()
        assert db_profile.first_name == "Bob"
        assert db_profile.last_name == "Smith"

    @pytest.mark.asyncio
    async def test_update_validation_failure(self, profile_service):
        """Test update with invalid data."""
        profile_data = {
            "firstName": "John",
            "lastName": "Doe",
            "email": "invalid-email"  # Invalid email
        }

        result = await profile_service.update(profile_data)

        assert result.success is False
        assert "validation failed" in result.message.lower()

    @pytest.mark.asyncio
    async def test_update_retry_logic(self, profile_service, db_session, monkeypatch):
        """Test retry logic with exponential backoff."""
        # Mock commit to fail twice, then succeed
        original_commit = db_session.commit
        commit_count = {"count": 0}

        def mock_commit():
            commit_count["count"] += 1
            if commit_count["count"] < 3:
                raise Exception("Simulated database error")
            return original_commit()

        monkeypatch.setattr(db_session, "commit", mock_commit)

        profile_data = {
            "firstName": "Retry",
            "lastName": "Test",
            "email": "retry@example.com"
        }

        result = await profile_service.update(profile_data, max_retries=3)

        assert result.success is True
        assert commit_count["count"] == 3  # Failed twice, succeeded third time


class TestProfileServiceValidation:
    """Test validate() method."""

    def test_validate_valid_profile(self, profile_service):
        """Test validation of valid profile data."""
        profile_data = {
            "firstName": "John",
            "lastName": "Doe",
            "email": "john@example.com",
            "phone": "+1234567890"
        }

        result = profile_service.validate(profile_data)

        assert result.isValid is True
        assert all(error is None for error in result.errors.values())

    def test_validate_invalid_email(self, profile_service):
        """Test validation of invalid email."""
        profile_data = {
            "firstName": "John",
            "lastName": "Doe",
            "email": "not-an-email",
        }

        result = profile_service.validate(profile_data)

        assert result.isValid is False
        assert result.errors["email"] is not None
        assert "valid email" in result.errors["email"].lower()

    def test_validate_invalid_phone(self, profile_service):
        """Test validation of invalid phone number."""
        profile_data = {
            "firstName": "John",
            "lastName": "Doe",
            "email": "john@example.com",
            "phone": "abc123"  # Invalid phone
        }

        result = profile_service.validate(profile_data)

        assert result.isValid is False
        assert result.errors["phone"] is not None
        assert "valid phone" in result.errors["phone"].lower()

    def test_validate_invalid_name_characters(self, profile_service):
        """Test validation of names with special characters."""
        profile_data = {
            "firstName": "John123",  # Numbers not allowed
            "lastName": "Doe@",  # Special characters not allowed
            "email": "john@example.com"
        }

        result = profile_service.validate(profile_data)

        assert result.isValid is False
        assert result.errors["firstName"] is not None
        assert result.errors["lastName"] is not None

    def test_validate_optional_fields(self, profile_service):
        """Test validation with optional fields omitted."""
        profile_data = {
            "firstName": "John",
            "lastName": "Doe",
            "email": "john@example.com",
            # phone omitted
        }

        result = profile_service.validate(profile_data)

        assert result.isValid is True

    def test_validate_hyphenated_names(self, profile_service):
        """Test validation allows hyphenated names."""
        profile_data = {
            "firstName": "Mary-Jane",
            "lastName": "O'Connor",
            "email": "mary@example.com"
        }

        result = profile_service.validate(profile_data)

        assert result.isValid is True


class TestProfileServiceClear:
    """Test clear() method."""

    @pytest.mark.asyncio
    async def test_clear_profile(self, profile_service, db_session):
        """Test clearing profile data."""
        # Create profile with data
        profile = db_session.query(UserProfile).filter(UserProfile.id == 1).first()
        profile.first_name = "John"
        profile.last_name = "Doe"
        profile.email = "john@example.com"
        profile.phone = "+1234567890"
        db_session.commit()

        await profile_service.clear()

        # Verify cleared
        db_profile = db_session.query(UserProfile).filter(UserProfile.id == 1).first()
        assert db_profile.first_name is None
        assert db_profile.last_name is None
        assert db_profile.email is None
        assert db_profile.phone is None
        assert db_profile.name == "Legal User"  # Reset to default


class TestProfileServiceGetExtended:
    """Test get_extended() method with caching."""

    @pytest.mark.asyncio
    async def test_get_extended_with_full_name(self, profile_service, db_session):
        """Test getting extended profile with computed fields."""
        profile = db_session.query(UserProfile).filter(UserProfile.id == 1).first()
        profile.first_name = "John"
        profile.last_name = "Doe"
        profile.email = "encrypted_john@example.com"
        db_session.commit()

        result = await profile_service.get_extended()

        assert result is not None
        assert result.fullName == "John Doe"
        assert result.initials == "JD"

    @pytest.mark.asyncio
    async def test_get_extended_with_first_name_only(self, profile_service, db_session):
        """Test extended profile with only first name."""
        profile = db_session.query(UserProfile).filter(UserProfile.id == 1).first()
        profile.first_name = "Alice"
        profile.email = "encrypted_alice@example.com"
        db_session.commit()

        result = await profile_service.get_extended()

        assert result is not None
        assert result.fullName == "Alice"
        assert result.initials == "A"

    @pytest.mark.asyncio
    async def test_get_extended_caching(self, profile_service, db_session):
        """Test that extended profile is cached."""
        profile = db_session.query(UserProfile).filter(UserProfile.id == 1).first()
        profile.first_name = "Bob"
        profile.last_name = "Smith"
        profile.email = "encrypted_bob@example.com"
        db_session.commit()

        # First call
        result1 = await profile_service.get_extended()

        # Change database (should not affect cached result)
        profile.first_name = "Changed"
        db_session.commit()

        # Second call (within cache duration)
        result2 = await profile_service.get_extended()

        assert result1.fullName == result2.fullName == "Bob Smith"

    @pytest.mark.asyncio
    async def test_get_extended_cache_invalidation(self, profile_service, db_session):
        """Test cache invalidation after update."""
        profile = db_session.query(UserProfile).filter(UserProfile.id == 1).first()
        profile.first_name = "Charlie"
        profile.last_name = "Brown"
        profile.email = "encrypted_charlie@example.com"
        db_session.commit()

        # Get cached result
        result1 = await profile_service.get_extended()
        assert result1.fullName == "Charlie Brown"

        # Update profile (should invalidate cache)
        await profile_service.update({"firstName": "Updated"})

        # Get new result (cache invalidated)
        result2 = await profile_service.get_extended()
        assert result2.fullName == "Updated Brown"

    @pytest.mark.asyncio
    async def test_get_extended_cache_expiration(self, profile_service, db_session, monkeypatch):
        """Test cache expiration after 5 seconds."""
        profile = db_session.query(UserProfile).filter(UserProfile.id == 1).first()
        profile.first_name = "David"
        profile.last_name = "Wilson"
        profile.email = "encrypted_david@example.com"
        db_session.commit()

        # First call
        result1 = await profile_service.get_extended()

        # Simulate time passing (mock time.time)
        original_time = time.time
        monkeypatch.setattr(time, "time", lambda: original_time() + 6)  # 6 seconds later

        # Change database
        profile.first_name = "NewName"
        db_session.commit()

        # Second call (cache expired)
        result2 = await profile_service.get_extended()

        assert result1.fullName == "David Wilson"
        assert result2.fullName == "NewName Wilson"


class TestProfileServiceConversions:
    """Test form_data_to_profile and profile_to_form_data methods."""

    def test_form_data_to_profile(self, profile_service):
        """Test converting form data to profile data."""
        form_data = ProfileFormData(
            firstName="  John  ",
            lastName="  Doe  ",
            email="  john@example.com  ",
            phone="  +1234567890  "
        )

        result = profile_service.form_data_to_profile(form_data)

        assert result.firstName == "John"  # Trimmed
        assert result.lastName == "Doe"
        assert result.email == "john@example.com"
        assert result.phone == "+1234567890"

    def test_profile_to_form_data_with_data(self, profile_service):
        """Test converting profile data to form data."""
        profile = UserProfileData(
            firstName="Jane",
            lastName="Smith",
            email="jane@example.com",
            phone="+9876543210"
        )

        result = profile_service.profile_to_form_data(profile)

        assert result.firstName == "Jane"
        assert result.lastName == "Smith"
        assert result.email == "jane@example.com"
        assert result.phone == "+9876543210"

    def test_profile_to_form_data_with_none(self, profile_service):
        """Test converting None to form data (defaults)."""
        result = profile_service.profile_to_form_data(None)

        assert result.firstName == ""
        assert result.lastName == ""
        assert result.email == ""
        assert result.phone == ""


class TestProfileServiceAuditLogging:
    """Test audit logging integration."""

    @pytest.mark.asyncio
    async def test_audit_logging_on_get(self, profile_service, db_session, audit_logger):
        """Test audit log on get operation."""
        profile = db_session.query(UserProfile).filter(UserProfile.id == 1).first()
        profile.first_name = "Test"
        profile.email = "encrypted_test@example.com"
        db_session.commit()

        await profile_service.get()

        # Verify audit log called
        assert audit_logger.log.called
        call_args = audit_logger.log.call_args[0][0]
        assert call_args["event_type"] == "profile.get"
        assert call_args["success"] is True

    @pytest.mark.asyncio
    async def test_audit_logging_on_update(self, profile_service, audit_logger):
        """Test audit log on update operation."""
        await profile_service.update({
            "firstName": "Audit",
            "lastName": "Test",
            "email": "audit@example.com"
        })

        # Verify audit log called
        assert audit_logger.log.called
        call_args = audit_logger.log.call_args[0][0]
        assert call_args["event_type"] == "profile.update"
        assert call_args["success"] is True

    @pytest.mark.asyncio
    async def test_audit_logging_on_clear(self, profile_service, audit_logger):
        """Test audit log on clear operation."""
        await profile_service.clear()

        # Verify audit log called
        assert audit_logger.log.called
        call_args = audit_logger.log.call_args[0][0]
        assert call_args["event_type"] == "profile.clear"
        assert call_args["success"] is True


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
